base/array.jl

```diff
@@ -2364,7 +2364,28 @@ function filter!(f, a::AbstractVector)
     return a
 end
 
-filter(f, a::Vector) = mapfilter(f, push!, a, empty(a))
+function filter!(f, a::Vector)
+    j = 1
+    for ai in a
+        @inbounds a[j] = ai
+        j = ifelse(f(ai), j+1, j)
+    end
+    deleteat!(a, j:length(a))
+    sizehint!(a, length(a))
+    a
+end
+
+function filter(f, a::AbstractVector{T}) where T
+    j = 1
```

Comment:
Need to limit to `Vector` with this. Also, can't this implementation just be
```
filter(f, a::Vector) = filter!(f, copy(a))
```
? How much difference is the `copy` vs undef-initialize?

Reply:
Copy vs undef-init: If only few elements of a large vector are selected, potentially large: The tail of the `undef` vector are mmaped by malloc and never faulted in. I measure
```
(p, N) = (0.01, 1000000)
  1.002 ms (1 allocation: 0 bytes)
  1.881 ms (3 allocations: 7.63 MiB)
(p, N) = (0.05, 1000000)
  1.046 ms (1 allocation: 0 bytes)
  1.924 ms (3 allocations: 7.63 MiB)
(p, N) = (0.15, 1000000)
  1.191 ms (1 allocation: 0 bytes)
  2.080 ms (3 allocations: 7.63 MiB)
(p, N) = (0.5, 1000000)
  1.582 ms (1 allocation: 0 bytes)
  2.457 ms (3 allocations: 7.63 MiB)
(p, N) = (0.85, 1000000)
  2.041 ms (1 allocation: 0 bytes)
  2.809 ms (3 allocations: 7.63 MiB)
(p, N) = (0.95, 1000000)
  1.163 ms (0 allocations: 0 bytes)
  2.011 ms (2 allocations: 7.63 MiB)
(p, N) = (0.99, 1000000)
  1.112 ms (0 allocations: 0 bytes)
  2.000 ms (2 allocations: 7.63 MiB)
```
If I skip the `sizehint!` and take the copy, I measure
```
(p, N) = (0.01, 1000000)
  977.451 μs (0 allocations: 0 bytes)
  1.852 ms (2 allocations: 7.63 MiB)
(p, N) = (0.05, 1000000)
  1.002 ms (0 allocations: 0 bytes)
  1.885 ms (2 allocations: 7.63 MiB)
(p, N) = (0.15, 1000000)
  1.028 ms (0 allocations: 0 bytes)
  1.899 ms (2 allocations: 7.63 MiB)
(p, N) = (0.5, 1000000)
  1.129 ms (0 allocations: 0 bytes)
  2.003 ms (2 allocations: 7.63 MiB)
(p, N) = (0.85, 1000000)
  1.178 ms (0 allocations: 0 bytes)
  2.074 ms (2 allocations: 7.63 MiB)
(p, N) = (0.95, 1000000)
  1.161 ms (0 allocations: 0 bytes)
  2.038 ms (2 allocations: 7.63 MiB)
(p, N) = (0.99, 1000000)
  1.120 ms (0 allocations: 0 bytes)
  1.992 ms (2 allocations: 7.63 MiB)
```
My tests were on archlinux with default config. This especially means that a large part of the time for `copy` is spent in the kernel pagefault handler (glibc malloc decides in its infinite wisdom to map the 8MB array with 4k pages without `madvise`ing the kernel that large pages are advisable, and archlinux default config decides in its infinite wisdom to respect userspace hints on pagesizes).

Reply:
Marking as resolved: I included limits on dispatch now, and using inplace filtering on a copy has significant overhead with rather insignificant gains in code complexity. Thanks for the review!